code/scripts/espurna_utils/ldscripts.py

```diff
@@ -22,3 +22,39 @@ def check_local_ld(target, source, env):
             env.Prepend(LIBPATH=[os.path.join(libpath_base, "latest")])
 
     env.AddPreAction(os.path.join("$BUILD_DIR", "firmware.elf"), check_local_ld)
+
+    # Bosch Sensortec Environment Cluster (BSEC) support
+    build_flags = env.ParseFlags(env['BUILD_FLAGS'])
+    espurna_flags = env.ParseFlags(env['ESPURNA_FLAGS'])
+
+    flags_with_value_list = [build_flag for build_flag in build_flags.get('CPPDEFINES') if type(build_flag) == list]
+    build_defines = {k: v for (k, v) in flags_with_value_list}
+
+    flags_with_value_list = [build_flag for build_flag in espurna_flags.get('CPPDEFINES') if type(build_flag) == list]
+    espurna_defines = {k: v for (k, v) in flags_with_value_list}
+
+    if build_defines.get('BME680_SUPPORT') == '1' or espurna_defines.get('BME680_SUPPORT') == '1':
+        ldscripts_inject_libpath_bme680(env)
+
+def ldscripts_inject_libpath_bme680(env):
+    platform = env.PioPlatform()
+    framework_dir = platform.get_package_dir("framework-arduinoespressif8266")
+    libpath_sdk = os.path.join(framework_dir, "tools", "sdk", "ld")
+    ldpath = os.path.join(libpath_sdk, "eagle.app.v6.common.ld")
+
+    # Due to the architecture of the ESP8266's memory and current size of the BSEC library,
+    # we need to modify the linker script and specifically define where the library should
+    # be placed in memory, otherwise the error 'section `.text1' will not fit in region
+    # `iram1_0_seg' will be returned.
+    pattern = re.compile(".*irom0.literal.*")
+    with open(ldpath, "r") as sources:
+        lines = sources.readlines()
+    with open(ldpath, "w") as sources:
+        for line in lines:
+            if pattern.match(line):
+                sources.write("    *libalgobsec.a:(.literal.* .text.*)\n")
+            sources.write(line)
+
+    # Load proprietary BSEC software of precise IAQ measurement.
+    env.Append(LIBPATH=["${PROJECT_LIBDEPS_DIR}/${PIOENV}/BSEC Software Library/src/esp8266"])
+    env.Append(LIBS=["algobsec"])
```

Comment:
While these two lines are probably useful, as the only other way is to add literal `-lalgobsec` into the cmdline flags in the .ini, patching .ld script like that is really hacky.

- maybe open an issue in the bosch repo to actually provide an .a file with correct sections? :)
- perhaps we can modify things ourselves. for example:
https://gist.github.com/mcspr/e61ee6162163d69f0806f0bd681033df
```
$ cd <some place with archive a from the repo>
$ ar x <archive>
$ # do the script
$ cd updated
$ ~/.platformio/packages/toolchain-xtensa/bin/xtensa-lx106-elf-ar cr libalgobsec.a *.o
$ mv libalgobsec.a <archive orginal place>
```

Plus, I don't think we detect flags in custom.h / hardware.h / sensor.h that way.

Reply:
Another option is creating a dedicated environment, but I would love if the experience for the end user is completely seamless and also unrelated to the base board / device.

Reply:
In either case, patching needs to always happen, b/c we can use headers. Travis fails exactly because of using custom.h and not providing ESPURNA_FLAGS or having any build_flags.
 
In theory PIO should detect library usage based on \#include definitions detection when using lib_ldf_mode=chain+ (or deep+), but it worked really bad last time I tried it, creating N^2 dependencies because of recursive includes :/

Reply:
Bosch has included many references to this problem across their documentation. I'm not sure if this is a case of 'we don't know better' (which I find hard, tbh) or simply because they don't have any incentive to fix it, but nonetheless if this can be changed on their side, I can submit a ticket.

Regarding your patching script, it seems like a very good alternative and less error-prone. Should we simply call script.sh or would you rather generate this dynamically every single time?

Reply:
Build configuration is Arduino IDE's weird side, in the perfect work it would be controlled by the user. Like, PIO actually allows this, both adding libraries and custom settings for the app and not through editing framework files.

Generation can happen once, a-la lwip build modification. .sh it's just something I came up on the go, this might as well be called exclusively through the python script - backup original .a, unpack .a, patch sections, pack it back up, replace .a + add some .DONE file

Reply:
re. as to really why...
https://github.com/esp8266/Arduino/blob/cc1cc0b2ce4825057f6a8bb0c6b1b82e17be313e/tools/sdk/ld/eagle.app.v6.common.ld.h#L138-L144

And notice now bsec files are named :) Perhaps there is even an easier way - just rename them! And pack back.

Reply:
That's pretty neat. So you're suggesting that simply renaming them to `*.cpp.o`  would do the trick.

Reply:
Yes. Probably .c.o though, but it should not matter (probably)